sklearn/preprocessing/_polynomial.py

```diff
@@ -393,18 +404,56 @@ def transform(self, X):
                 )
             if self._min_degree <= 1 and self._max_degree > 0:
                 to_stack.append(X)
+
+            # Covnert current stack to valid indices/indptr types
+            for mat in to_stack:
+                mat.indices = mat.indices.astype(index_t, copy=False)
+                mat.indptr = mat.indptr.astype(index_t, copy=False)
             for deg in range(max(2, self._min_degree), self._max_degree + 1):
-                Xp_next = _csr_polynomial_expansion(
-                    X.data, X.indices, X.indptr, X.shape[1], self.interaction_only, deg
+                # Count how many nonzero elements the expanded matrix will contain.
+                total_nnz = sum(
+                    _calc_expanded_dimensionality(
+                        X.indptr[row_i + 1] - X.indptr[row_i],
+                        self.interaction_only,
+                        deg,
+                    )
+                    for row_i in range(X.indptr.shape[0] - 1)
+                )
+                expanded_d = _calc_expanded_dimensionality(
+                    X.shape[1], self.interaction_only, deg
                 )
-                if Xp_next is None:
+
+                if expanded_d == 0:
                     break
-                to_stack.append(Xp_next)
+                assert expanded_d > 0
```

Comment:
As a safety, we should also check that if `total_nnz > max_int32` then we should use `index_t = np.int64` for the subsequent index arrays.

Reply:
This was originally considered in https://github.com/scikit-learn/scikit-learn/pull/23731#discussion_r907128501 which was marked as resolved without being fully resolved.